src/OrleansRuntime/Placement/IPlacementContext.cs

```diff
@@ -54,22 +56,32 @@ internal interface IPlacementContext
         /// <returns></returns>
         bool TryGetActivationData(ActivationId id, out ActivationData activationData);
 
-        void GetGrainTypeInfo(int typeCode, out string grainClass, out PlacementStrategy placement, string genericArguments = null);
+        void GetGrainTypeInfo(int typeCode, out string grainClass, out PlacementStrategy placement, out MultiClusterRegistrationStrategy strategy, string genericArguments = null);
     }
 
     internal static class PlacementContextExtensions
     {
-        public static Task<List<ActivationAddress>> Lookup(this IPlacementContext @this, GrainId grainId)
+        public static async Task<List<ActivationAddress>> Lookup(this IPlacementContext @this, GrainId grainId)
         {
             List<ActivationAddress> l;
-            return @this.FastLookup(grainId, out l) ? Task.FromResult(l) : @this.FullLookup(grainId);
+            if (@this.FastLookup(grainId, out l))
+            {
+                return l;
+            }
+            else
+            {
+                var result = await @this.FullLookup(grainId);
```

Comment:
There is now an extra `await` being done on every call to that func, even if we exit via `FastLookup`. This func is called on EVERY msg (cache lookup), so it is worth optimizing.
You can do the same trick @ReubenBond did in #1074: leave the outer function NON async.
The slow path - in the else clause - do not put await. Instead call another helper function WRAPPER which internally we call `FullLookup` and do `await` on it and `Lookup` will just return what WRAPPER  returns. As a result, you will avoid `await` state machine allocation on the fast path. Not worth doing this everywhere, but worth here I think.